core/trino-main/src/main/java/io/trino/execution/scheduler/faulttolerant/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -492,13 +495,152 @@ public void failTaskRemotely(TaskId taskId, Throwable failureCause)
+        {
+            if (!log.isDebugEnabled()) {
+                return;
+            }
+            log.debug("SqlStages:");
+            stages.forEach((stageId, stage) -> {
+                log.debug("SqlStage %s: %s", stageId, stage);
+            });
+        }
+    }
+
+    private static class EventDebugInfos
+    {
+        private static final String GLOBAL_EVENTS_BUCKET = "GLOBAL";
+        private static final EventListener<String> GET_BUCKET_LISTENER = new EventListener<>()
+        {
+            @Override
+            public String onRemoteTaskEvent(RemoteTaskEvent event)
+            {
+                return "task_" + event.getTaskStatus().getTaskId().getStageId().toString();
+            }
+
+            @Override
+            public String onRemoteTaskExchangeUpdatedSinkAcquired(RemoteTaskExchangeUpdatedSinkAcquired event)
+            {
+                return "task_" + event.getTaskId().getStageId().toString();
+            }
+
+            @Override
+            public String onStageEvent(StageEvent event)
+            {
+                return event.getStageId().toString();
+            }
+
+            @Override
+            public String onEvent(Event event)
+            {
+                return GLOBAL_EVENTS_BUCKET;
+            }
+        };
+
+        private final String queryId;
+        private final int eventsPerBucket;
+        private long eventsCounter;
+        private long filteredEventsCounter;
+
+        private volatile SoftReference<ListMultimap<String, String>> eventsDebugInfosReference;
+
+        private EventDebugInfos(String queryId, int eventsPerBucket)
+        {
+            this.queryId = requireNonNull(queryId, "queryId is null");
+            this.eventsPerBucket = eventsPerBucket;
+            eventsDebugInfosReference = new SoftReference<>(LinkedListMultimap.create());
+        }
+
+        private boolean add(Event event)
+        {
+            ListMultimap<String, String> eventsDebugInfos = getEventsDebugInfos();
+            String bucket = getBucket(event);
+            Optional<String> debugInfo = getFullDebugInfo(eventsCounter, event);
+            eventsCounter++;
+            if (debugInfo.isEmpty()) {
+                filteredEventsCounter++;
+                return false;
+            }
+            eventsDebugInfos.put(bucket, debugInfo.get());
+            List<String> bucketDebugInfos = eventsDebugInfos.get(bucket);
+            if (bucketDebugInfos.size() > eventsPerBucket) {
+                Iterator<String> iterator = bucketDebugInfos.iterator();
+                iterator.next();
+                iterator.remove();
+            }
+            return true;
+        }
+
+        private ListMultimap<String, String> getEventsDebugInfos()
+        {
+            ListMultimap<String, String> eventsDebugInfos = eventsDebugInfosReference.get();
+            if (eventsDebugInfos == null) {
+                log.info("eventsDebugInfos for %s has been cleared", queryId);
```

Comment:
debug level

debug infos are useless unless debug is enabled
plus, this has potential of being logged many times, can flood